proposals/index-range-changes.md

```diff
@@ -0,0 +1,297 @@
+aren't any examples of slicing to look to for an example. Types like `List<T>`, `ArraySegment<T>`, `SortedList<T>`
+would've been ideal for slicing but the concept didn't exist when types were added. 
+
+Thus, `Slice` being the sole example, it was chosen as the name.
+
+### Index target type conversion
+Another way to view the `Index` transformation in an indexer expression is as a target type conversion. Instead
+of binding as if there is a member of the form `return_type this[Index]`, the language instead assigns a target
+typed conversion to `int`. 
+
+This concept could be generalized to all member access on Countable types. Whenever an expression with type `Index` is 
+used as an argument to an instance member invocation and the receiver is Countable then the expression will have a 
+target type conversion to `int`. The member invocations applicable for this conversion include methods, indexers, 
+properties, extension methods, etc ... Only constructors are excluded as they have no receiver. 
+
+The target type conversion will be implemented as follows for any expression which has a type of `Index`. For
+discussion purposes lets use the example of `receiver[expr]`:
+
+- When `expr` is of the form `^expr2` and the type of `expr2` is `int`, it will be translated to 
+`receiver.Length - expr2`.
+- Otherwise, it will be translated as `expr.GetOffset(receiver.Length)`.
+
+The `receiver` and `Length` expressions will be spilled as appropriate to ensure any side effects are only executed
+once. For example:
+
+``` csharp
+class Collection {
+    private int[] _array = new[] { 1, 2, 3 };
+
+    int Length {
+        get {
+            Console.Write("Length ");
+            return _array.Length;
+        }
+    }
+
+    int GetAt(int index) => _array[index];
+}
+
+class SideEffect {
+    Collection Get() {
+        Console.Write("Get ");
+        return new Collection();
+    }
+
+    void Use() { 
+        int i = Get().GetAt(^1);
+        Console.WriteLine(i);
+    }
+}
+```
+
+This code will print "Get Length 3". 
+
+This feature would be beneficial to any member which had a parameter that represented an index. For example 
+`List<T>.InsertAt`. This also has the potential for confusion as the language can't give any guidance as to whether
+or not an expression is meant for indexing. All it can do is convert any `Index` expression to `int` when invoking
+a member on a Countable type. 
+
+Restrictions:
+
+- This conversion is only applicable when the expression with type `Index` is directly an argument to the member. It 
+would not apply to any nested expressions.
+
+## Related Issues
+- https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/ranges.cs
+- https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/ranges.md
+
+## Design Meetings
+- https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-04-01.md
```

Comment:
@tarekgh, @terrajobst, we should review all of the Index/Range-based members we added to decide what to remove and what to add, if anything.  At a minimum we should remove all of the Index/Range-based indexers that were added.

Reply:
https://github.com/Microsoft/dotnet-blog/blob/master/2019/03-Mar/dotnet-core3-preview3/dotnetcore3-preview3.md has all stuff we exposed in preview 3. before then we added only the Span indexers.